### PR TITLE
AIP-82 Create references between triggers and assets

### LINKED ISSUE #42509: AIP-82. Create migration to add relation between assets and triggers
### Description

_No response_

### Use case/motivation

_No response_

### Related issues

_No response_

### Are you willing to submit a PR?

- [x] Yes I am willing to submit a PR!

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by vincbeck**
Resolves #42509.

This migration creates a table to store relations between triggers and assets. The goal is a trigger can update an asset whenever an external change is detected.

For more details on AIP-82, please [read the AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-82+External+event+driven+scheduling+in+Airflow).

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Review by dstandish — COMMENTED** [review]
Hey, it's not really clear exactly what we're doing with this relationship and therefore it's kindof hard to review -- we don't know how it will be used, so it's not clear if this is the right structure as opposed to say a many to many relationship.  

Why do we need a key between the trigger and the asset in order for the trigger to do something with the asset or in response to an asset update?  E.g. couldn't it just take an asset id as kwargs to the trigger obj?

**Comment by vincbeck** [comment]
> Hey, it's not really clear exactly what we're doing with this relationship and therefore it's kindof hard to review -- we don't know how it will be used, so it's not clear if this is the right structure as opposed to say a many to many relationship.
> 
> Why do we need a key between the trigger and the asset in order for the trigger to do something with the asset or in response to an asset update? E.g. couldn't it just take an asset id as kwargs to the trigger obj?

The goal here is to create a table to store references between assets and triggers. The relation here is many to many (hence the table). Or maybe I did something wrong and the relation is not many to many? Nonetheless, the relation SHOULD be many to many :)

Why do we want to associate assets and triggers? The ultimate goal is to add `watchers` to assets. These watchers are triggers and whenever the trigger is triggered ... it updates the asset. See example below:

```
trigger = SqsSensorTrigger(sqs_queue="<my_queue>")
asset = Asset("<my_queue>", watchers=[trigger])
```

Whenever the trigger `trigger` is triggered, the asset `asset` will be sent an event 

> E.g. couldn't it just take an asset id as kwargs to the trigger obj?

I am not sure I understand this one :)

**Comment by dstandish** [comment]
> > E.g. couldn't it just take an asset id as kwargs to the trigger obj?
> 
> I am not sure I understand this one :)

I was saying like we create a trigger like `MyTrigger(asset_id: str, other_stuff: Any)`

But I see now from your explanation and example that you are looking to change triggers so they are not specific to a TI, but rather that they may be long lived, not associated with a TI, and can be instead associated to assets.  And one asset can be associated to many triggers.  And one trigger can be associated with many assets.

This is a big one 😅

**Comment by vincbeck** [comment]
> > > E.g. couldn't it just take an asset id as kwargs to the trigger obj?
> > 
> > 
> > I am not sure I understand this one :)
> 
> I was saying like we create a trigger like `MyTrigger(asset_id: str, other_stuff: Any)`
> 
> But I see now from your explanation and example that you are looking to change triggers so they are not specific to a TI, but rather that they may be long lived, not associated with a TI, and can be instead associated to assets. And one asset can be associated to many triggers. And one trigger can be associated with many assets.
> 
> This is a big one 😅

Haha, yeah :) This is AIP-82 :) First cornerstone :D But I am happy you understood exactly what I am trying to do

**Review by dstandish — APPROVED** [review]

File: airflow/models/asset.py [hunk 1]
```diff
@@ -284,6 +285,46 @@ def for_asset(cls, asset: AssetModel) -> AssetActive:
...
+            ["asset.id"],
+            ondelete="CASCADE",
+        ),
+    )
+
+    def __eq__(self, other):
+        if isinstance(other, self.__class__):
+            return self.trigger_id == other.trigger_id and self.asset_id == other.asset_id
+        else:
+            return NotImplemented
```
Comment (Lee-W) [inline airflow/models/asset.py]:
```suggestion
        return NotImplemented
```

nit

File: airflow/models/asset.py [hunk 2]
```diff
@@ -284,6 +285,46 @@ def for_asset(cls, asset: AssetModel) -> AssetActive:
...
+
+    def __hash__(self):
```
Comment (Lee-W) [inline airflow/models/asset.py]: May I know what is this for?

File: airflow/models/asset.py [hunk 3]
```diff
@@ -284,6 +285,46 @@ def for_asset(cls, asset: AssetModel) -> AssetActive:
...
+        return hash(self.__mapper__.primary_key)
+
+    def __repr__(self):
+        args = []
+        for attr in [x.name for x in self.__mapper__.primary_key]:
+            args.append(f"{attr}={getattr(self, attr)!r}")
```
Comment (Lee-W) [inline airflow/models/asset.py]:
```suggestion
        args = [
            f"{x.name}={getattr(self, x.name)!r}"
            for x in self.__mapper__.primary_key
        ]
```

File: airflow/models/asset.py (hunk 2 above)
Comment (vincbeck) [reply airflow/models/asset.py]: I was following the other models in that class and they all implement these methods, happy to remove them if you think they are not necessary

File: airflow/models/asset.py (hunk 1 above)
Comment (vincbeck) [reply airflow/models/asset.py]: Sure

File: airflow/models/asset.py (hunk 3 above)
Comment (vincbeck) [reply airflow/models/asset.py]: Sure

**Review by o-nikolas — APPROVED** [review]

File: airflow/models/asset.py (hunk 2 above)
Comment (Lee-W) [reply airflow/models/asset.py]: Thanks! I was suggested to add it when we needed it back to the time DatasetAlias was introduced. It's easier for us to add new things then removing ones.

**Review by Lee-W — APPROVED** [review]

**Pull request merged**